### ISSUE TITLE
NS with final dot makes the test die

### ISSUE THREAD

**Issue by blacksponge** [zonemaster/zonemaster-backend#854]
When starting a undelegated test with a NS that ends with a final dot, the test will die with error

```
Each key of argument $href must omit the trailing dot, or it must be a single dot at /usr/local/share/perl5/Zonemaster/Backend/TestAgent.pm line 208.
```

The NSs domains  should maybe be normalized before being passed to the engine.

**Comment by mattias-p** [comment]
One idea is to restrict the RPCAPI so that it only accepts domain names without a trailing dot. This way we'd have a single normalization form to deal with internally, and the normalization form used in both inputs and outputs would be the same. I believe this would result in fewer surprises for our users (including ourselves). To also make it ergonomic for clients we could have a special method in the RPCAPI whose sole purpose is domain name normalization.

If trailing dots was the only normalization aspect for domain names I wouldn't suggest this approach. But there's also character cases and a number of forms for internationalized domain names. Lump all these together and I think it would be nice to have restrictive business logic methods that only accept domain names in a single normalization form, and a special normalization method that accepts any reasonable combination of these forms and returns one ASCII domain name (that never contains any U-labels) and one Unicode domain name (that never contains any A-labels). Both in lower-case and both without a trailing dot.

I'm sure there's another issue suggesting this, but now I can't find it.

**Comment by matsduf** [comment]
There is a PR for Basic00 covering all this, zonemaster/zonemaster#942. The suggestion from @mattias-p above is that part of suggested Basic00 is implemented in the end clients, GUI and (I assume) CLI and that there is a RPCAPI method doing the job. Should that then be an Engine method so that the same code could be shared with CLI to ensure consistent implementation?

**Comment by pnax** [comment]
Fixed with #983
